src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,742 @@
+      const rowLength =
+        (checkSize.width / kSizedTextureFormatInfo[format].blockWidth!) *
+        kSizedTextureFormatInfo[format].bytesPerBlock!;
+      this.expectContents(buffer, expected.slice(rowOffset, rowOffset + rowLength), rowOffset);
+    }
+  }
+
+  getFullData(
+    { texture, mipLevel }: { texture: GPUTexture; mipLevel: number | undefined },
+    fullTextureCopyLayout: TextureCopyLayout
+  ): GPUBuffer {
+    const { mipSize, byteLength, bytesPerRow, rowsPerImage } = fullTextureCopyLayout;
+    const buffer = this.device.createBuffer({
+      size: align(byteLength, 4), // this is necessary because we need to copy and map data from this buffer
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    const encoder = this.device.createCommandEncoder();
+    encoder.copyTextureToBuffer(
+      { texture, mipLevel },
+      { buffer, bytesPerRow, rowsPerImage },
+      mipSize
+    );
+    this.device.defaultQueue.submit([encoder.finish()]);
+
+    return buffer;
+  }
+
+  updateFullData(
+    fullTextureCopyLayout: TextureCopyLayout,
+    texturePartialDataLayout: GPUTextureDataLayout,
+    copySize: GPUExtent3DDict,
+    origin: Required<GPUOrigin3DDict>,
+    format: SizedTextureFormat,
+    fullData: Uint8Array,
+    partialData: Uint8Array
+  ): void {
+    for (const texel of this.iterateBlockRows(copySize, origin, format)) {
+      const partialDataOffset = this.getTexelOffsetInBytes(
+        texturePartialDataLayout,
+        format,
+        texel,
+        origin
+      );
+      const fullDataOffset = this.getTexelOffsetInBytes(
+        {
+          bytesPerRow: fullTextureCopyLayout.bytesPerRow,
+          rowsPerImage: fullTextureCopyLayout.rowsPerImage,
+          offset: 0,
+        },
+        format,
+        texel
+      );
+      const rowLength =
+        (copySize.width / kSizedTextureFormatInfo[format].blockWidth!) *
+        kSizedTextureFormatInfo[format].bytesPerBlock!;
+      for (let b = 0; b < rowLength; ++b) {
+        fullData[fullDataOffset + b] = partialData[partialDataOffset + b];
+      }
+    }
+  }
+
+  fullCheck(
+    { texture, mipLevel, origin }: TextureCopyViewWithRequiredOrigin,
+    fullTextureCopyLayout: TextureCopyLayout,
+    texturePartialDataLayout: GPUTextureDataLayout,
+    copySize: GPUExtent3DDict,
+    format: SizedTextureFormat,
+    fullData: GPUBuffer,
+    partialData: Uint8Array
+  ): void {
+    const { mipSize, bytesPerRow, rowsPerImage, byteLength } = fullTextureCopyLayout;
+    const { dst, begin, end } = this.createAlignedCopyForMapRead(fullData, byteLength, 0);
+
+    // We add an eventual async expectation which will update the full data and then add
+    // other eventual async expectations to ensure it will be correct.
+    this.eventualAsyncExpectation(async () => {
+      await dst.mapAsync(GPUMapMode.READ);
+      const actual = new Uint8Array(dst.getMappedRange()).slice(begin, end);
+      this.updateFullData(
```

Comment:
updateFullData could be done before entering the eventual expectation. In general we should minimize the amount of work done in eventual expectations, and check against known data as much as possible.